e5e.go

```diff
@@ -1,261 +1,389 @@
-//
-//goland:noinspection GoUnusedExportedFunction
-func Start(entrypoints interface{}) {
-	exitCode, err := start(entrypoints)
-	if err != nil {
-		panic(err)
-	}
-	os.Exit(exitCode)
-}
+// Options contains all the runtime options that determine the behaviour of the [Mux].
+// It is usually read at runtime, but can be overridden for testing.
+type Options struct {
+	// If set to true, write the metadata at startup and stop the application.
+	WriteMetadata bool
 
-// Internal implementation of the Start logic which returns an exit code and any available error object.
-func start(entrypoints interface{}) (int, error) {
-	executionInstance := execution{}
+	// The name of the entrypoint that is executed on incoming events.
+	Entrypoint string
 
-	// Init execution instance
-	if err := executionInstance.init(entrypoints); err != nil {
-		return -1, fmt.Errorf("go-e5e: %w", err)
-	}
+	// The termination sequence that should be written on shutdown.
+	DaemonExecutionTerminationSequence string
 
-	// Start execution loop
-	for {
-		if err := executionInstance.execute(); err != nil {
-			return -1, fmt.Errorf("go-e5e: %w", err)
-		}
+	// The execution sequence that separates generic output on [os.Stdout] from the encoded responses.
+	StdoutExecutionSequence string
 
-		// In case this is a single execution exit the loop
-		if !executionInstance.keepalive {
-			break
-		}
+	// If set to true, the application is kept alive after the first execution and responds to ping events.
+	KeepAlive bool
+}
 
-		// Print execution termination signals
-		_, _ = fmt.Fprint(os.Stdout, executionInstance.daemonExecutionTermination)
-		_, _ = fmt.Fprint(os.Stderr, executionInstance.daemonExecutionTermination)
-	}
+// Mux defines a container for entrypoints and routes the requests for the given entrypoint
+// to the dedicated handlers.
+type Mux struct {
+	handlers    map[string]Handler
+	stdinReader *bufio.Scanner
+
+	// If set, the Mux uses the given reader for its input.
+	// Otherwise, [os.Stdin] is used.
+	Stdin io.Reader
+
+	// If set, the Mux uses the given writer for its output.
+	// Otherwise, [os.Stdout] is used.
+	Stdout io.Writer
 
-	return 0, nil
+	// If set, the Mux uses the given writer for its output.
+	// Otherwise, [os.Stderr] is used.
+	Stderr io.Writer
+
+	// If set, the options are not read during the first invocation.
+	// Otherwise it is set to the parsed result of [os.Args].
+	Options *Options
 }
 
-// Populates the execution struct with values given as command line arguments and fetches the required entrypoint. Will
-// return an error on any invalid condition.
-func (e *execution) init(entrypoints interface{}) error {
-	// Check number of arguments:
-	// binary name, entrypoint, os.Stdout termination, keepalive enabled, daemon execution termination
-	if argCount := len(os.Args); argCount != 5 {
-		return fmt.Errorf("invalid number of process arguments: %d", argCount)
+// NewMux creates a new mux.
+func NewMux() *Mux {
```

Comment:
I'm not sure there even is a reason to have multiple `Mux` instances. Could we just replace that with a private global one and use the function handler add functions to append there?

Reply:
The only reason for a dedicated struct is the fact that global variables are kinda an anti-pattern. Furthermore it allows us to encapsulate the std* streams into the mux, which allows much easier mocking in the tests.